api/client/client.go

```diff
@@ -2170,3 +2173,79 @@ func (c *Client) GenerateCertAuthorityCRL(ctx context.Context, req *proto.CertAu
 	resp, err := c.grpc.GenerateCertAuthorityCRL(ctx, req)
 	return resp, trail.FromGRPC(err)
 }
+
+// ListResources returns a paginated list of nodes that the user has access to.
+// nextKey can be used as startKey in another call to ListResources to retrieve
+// the next page.
+// It will return a `trace.LimitExceeded` error if the page exceeds gRPC max
+// message size.
+func (c *Client) ListResources(ctx context.Context, req proto.ListResourcesRequest) ([]types.Resource, string, error) {
+	if req.Namespace == "" {
+		return nil, "", trace.BadParameter("missing parameter namespace")
+	}
+	if req.Limit <= 0 {
+		return nil, "", trace.BadParameter("nonpositive parameter limit")
+	}
```

Comment:
We usually implement a `CheckAndSetDefaults` function that validates the struct. For proto structs you can place it in types.go for example (https://github.com/gravitational/teleport/blob/v8.0.0/api/client/proto/types.go).